File: stations/_archive/hook-observatory-py/handlers/plan_impl_gate.py
"""
Plan-to-Implementation gate — detect ExitPlanMode and inject context reminder.

Handles two events:
  - PostToolUse (ExitPlanMode): Write marker file with plan path + timestamp
  - UserPromptSubmit: If marker exists, inject reminder to save plan to memory
    and consider starting a new session. One-shot: marker deleted after use.

Marker auto-expires after 1 hour to prevent stale triggers.
"""

from __future__ import annotations

import json
import os
import time

from .base import ALLOW, HookResult, text_result

_MARKER_DIR = os.path.join(os.path.expanduser("~"), ".hook-observatory", "markers")
_MARKER_PREFIX = ".plan-approved-"
_MARKER_TTL_SECONDS = 3600  # 1 hour


def _marker_path() -> str:
    session_id = os.environ.get("CLAUDE_SESSION_ID", "unknown")
    return os.path.join(_MARKER_DIR, f"{_MARKER_PREFIX}{session_id}")


def handle(event_type: str, tool_name: str, tool_input: dict, raw_input: str) -> HookResult:
    if event_type == "PostToolUse" and tool_name == "ExitPlanMode":
        return _on_exit_plan_mode(tool_input)

    if event_type == "UserPromptSubmit":
        return _on_user_prompt()

    return ALLOW


def _on_exit_plan_mode(tool_input: dict) -> HookResult:
    """Write marker file when exiting plan mode."""
    marker = _marker_path()
    try:
        os.makedirs(os.path.dirname(marker), exist_ok=True)
        data = {
            "timestamp": time.time(),
            "plan_path": tool_input.get("plan_path", ""),
        }
        with open(marker, "w") as f:
            json.dump(data, f)
    except OSError:
        pass  # fail-open: don't block on write failure
    return ALLOW


def _on_user_prompt() -> HookResult:
    """Check for plan-approved marker and inject reminder if found."""
    marker = _marker_path()

    try:
        if not os.path.exists(marker):
            return ALLOW

        with open(marker) as f:
            data = json.load(f)

        # Check TTL — ignore stale markers
        ts = data.get("timestamp", 0)
        if time.time() - ts > _MARKER_TTL_SECONDS:
            os.unlink(marker)
            return ALLOW

        # One-shot: delete marker before injecting
        os.unlink(marker)

        plan_path = data.get("plan_path", "")
        path_hint = f"\n- Plan file: `{plan_path}`" if plan_path else ""

        return text_result(
            f"## Plan-to-Impl Gate\n"
            f"剛完成計畫階段 (plan mode 消耗了大量 context)。建議:\n"
            f"1. 將計畫中的關鍵決策存到 memory (避免 compact 後遺失)\n"
            f"2. 考慮開新 session 以乾淨 context 開始實作{path_hint}"
        )
    except (OSError, json.JSONDecodeError, KeyError):
        # fail-open
        return ALLOW

File: internal/handlers/plan_impl_gate.go
package handlers

import (
	"encoding/json"
	"os"
	"path/filepath"
	"strings"
	"time"

	"github.com/joneshong/hook-observatory/internal/core"
)

func init() {
	core.Register("PostToolUse", core.Entry{
		Matcher:    "",
		Handler:    planImplGatePostToolUse,
		Critical:   false,
		ModuleName: "plan_impl_gate",
	})
	core.Register("UserPromptSubmit", core.Entry{
		Matcher:    "",
		Handler:    planImplGateUserPrompt,
		Critical:   false,
		ModuleName: "plan_impl_gate",
	})
}

const (
	pigMarkerPrefix = ".plan-approved-"
	pigMarkerTTL    = time.Hour
)

// pigMarkerDir returns the directory where plan-approved markers are stored.
func pigMarkerDir() string {
	home, _ := os.UserHomeDir()
	return filepath.Join(home, ".hook-observatory", "markers")
}

// pigMarkerPath returns the full path to the session's plan-approved marker.
func pigMarkerPath() string {
	sid := os.Getenv("CLAUDE_SESSION_ID")
	if sid == "" {
		sid = "unknown"
	}
	return filepath.Join(pigMarkerDir(), pigMarkerPrefix+sid)
}

// planImplGatePostToolUse handles the PostToolUse+ExitPlanMode event.
// Writes a marker file with timestamp and plan_path.
func planImplGatePostToolUse(eventType, toolName string, toolInput map[string]any, _ string) core.HookResult {
	if eventType != "PostToolUse" || toolName != "ExitPlanMode" {
		return core.Allow()
	}
	return pigOnExitPlanMode(toolInput)
}

// planImplGateUserPrompt handles the UserPromptSubmit event.
// If a plan-approved marker exists and is fresh, injects a reminder.
func planImplGateUserPrompt(eventType, _ string, _ map[string]any, _ string) core.HookResult {
	if eventType != "UserPromptSubmit" {
		return core.Allow()
	}
	return pigOnUserPrompt()
}

// ---------------------------------------------------------------------------
// Internal logic
// ---------------------------------------------------------------------------

type pigMarkerData struct {
	Timestamp float64 `json:"timestamp"`
	PlanPath  string  `json:"plan_path"`
}

func pigOnExitPlanMode(toolInput map[string]any) core.HookResult {
	marker := pigMarkerPath()
	if err := os.MkdirAll(filepath.Dir(marker), 0o755); err != nil {
		return core.Allow() // fail-open
	}

	planPath, _ := toolInput["plan_path"].(string)
	data := pigMarkerData{
		Timestamp: float64(time.Now().UnixNano()) / 1e9,
		PlanPath:  planPath,
	}

	raw, err := json.Marshal(data)
	if err != nil {
		return core.Allow()
	}
	_ = os.WriteFile(marker, raw, 0o644) // fail-open
	return core.Allow()
}

func pigOnUserPrompt() core.HookResult {
	marker := pigMarkerPath()

	raw, err := os.ReadFile(marker)
	if err != nil {
		return core.Allow() // marker doesn't exist
	}

	var data pigMarkerData
	if err := json.Unmarshal(raw, &data); err != nil {
		return core.Allow()
	}

	// Check TTL
	age := time.Since(time.Unix(int64(data.Timestamp), int64((data.Timestamp-float64(int64(data.Timestamp)))*1e9)))
	if age > pigMarkerTTL {
		_ = os.Remove(marker)
		return core.Allow()
	}

	// One-shot: delete marker before injecting reminder
	_ = os.Remove(marker)

	var pathHint string
	if data.PlanPath != "" {
		pathHint = "\n- Plan file: `" + data.PlanPath + "`"
	}

	return core.TextResult(
		"## Plan-to-Impl Gate\n" +
			"剛完成計畫階段 (plan mode 消耗了大量 context)。建議:\n" +
			"1. 將計畫中的關鍵決策存到 memory (避免 compact 後遺失)\n" +
			"2. 考慮開新 session 以乾淨 context 開始實作" + pathHint,
	)
}

// ---------------------------------------------------------------------------
// Ensure strings import is used (pigMarkerPrefix uses it indirectly — guard)
// ---------------------------------------------------------------------------

var _ = strings.TrimSpace // keep import live if refactored
